### ISSUE TITLE
Missing `overridable` only blows up during sorbet-runtime

### ISSUE THREAD

**Issue by johnkchow** [sorbet/sorbet#2746]
#### Input

[→ View on sorbet.run](https://sorbet.run/#%23%20typed%3A%20strict%0A%0Aclass%20A%0A%20%20extend%20T%3A%3ASig%0A%0A%20%20sig%20%7Boverridable.void%7D%0A%20%20def%20foo%3B%20end%0Aend%0A%0Aclass%20B%20%3C%20A%0A%20%20extend%20T%3A%3ASig%0A%0A%20%20sig%20%7Boverride.void%7D%0A%20%20def%20foo%3B%20end%0Aend%0A%0Aclass%20C%20%3C%20B%0A%20%20extend%20T%3A%3ASig%0A%0A%20%20%23%20NOTE%3A%20Here%20we're%20missing%20%60override%60%0A%20%20sig%20%7Bvoid%7D%0A%20%20def%20foo%0A%20%20end%0Aend%0A%0AC.new.foo%20%23%20this%20will%20blow%20up%20in%20prod)

```ruby
# typed: strict

class A
  extend T::Sig

  sig {overridable.void}
  def foo; end
end

class B < A
  extend T::Sig

  sig {override.void}
  def foo; end
end

class C < B
  extend T::Sig

  # NOTE: Here we're missing `override`
  sig {void}
  def foo
  end
end

C.new.foo # this will blow up in prod
```

#### Observed output

from `srb tc`
```
No errors! Great job.
```

from `sorbet-runtime`:

```
RuntimeError: You must use `.override` when overriding the existing method `foo`.
  Parent definition: B at /var/folders/gq/vz0kmn3d215bm2fztcwdk9040000gn/T/foo.rb:14
  Child definition:  C at /var/folders/gq/vz0kmn3d215bm2fztcwdk9040000gn/T/foo.rb:22
from /Users/me/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/sorbet-runtime-0.5.5410/lib/types/private/methods/signature_validation.rb:70:in `validate_override_mode'
```

#### Expected behavior

The `srb tc` should also show the same error msg as the runtime.

**Comment by jez** [comment]
Another example:

```ruby
# typed: true
require 'sorbet-runtime'
class Parent
  extend T::Sig
  sig {returns(Integer)}
  def foo
    0
  end
end
class Child < Parent
  extend T::Sig
  sig {returns(String)}
  def foo
    ''
  end
end
p Parent.new.foo
p Child.new.foo
```

No static errors, but raises at runtime:

```
❯ ruby foo.rb
0
Traceback (most recent call last):
        9: from foo.rb:19:in `<main>'
        8: from /Users/jez/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/sorbet-runtime-0.5.5937/lib/types/private/methods/_methods.rb:209:in `block in _on_method_added'
        7: from /Users/jez/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/sorbet-runtime-0.5.5937/lib/types/private/methods/_methods.rb:366:in `maybe_run_sig_block_for_method'
        6: from /Users/jez/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/sorbet-runtime-0.5.5937/lib/types/private/methods/_methods.rb:370:in `maybe_run_sig_block_for_key'
        5: from /Users/jez/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/sorbet-runtime-0.5.5937/lib/types/private/methods/_methods.rb:390:in `run_sig_block_for_key'
        4: from /Users/jez/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/sorbet-runtime-0.5.5937/lib/types/private/methods/_methods.rb:200:in `block in _on_method_added'
        3: from /Users/jez/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/sorbet-runtime-0.5.5937/lib/types/private/methods/_methods.rb:295:in `run_sig'
        2: from /Users/jez/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/sorbet-runtime-0.5.5937/lib/types/private/methods/_methods.rb:334:in `build_sig'
        1: from /Users/jez/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/sorbet-runtime-0.5.5937/lib/types/private/methods/signature_validation.rb:43:in `validate'
/Users/jez/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/sorbet-runtime-0.5.5937/lib/types/private/methods/signature_validation.rb:208:in `validate_override_types': Incompatible return type in override of method `foo`: (RuntimeError)
* Base: `Integer` (in Parent at foo.rb:6)
* Override: `String` (in Child at foo.rb:13)
(The types must be covariant.)
```

We already have some precedent for these kinds of errors statically. In the example above, explictily marking parent/child as abstract/overridable will emit the errors. But as gdritter says for abstract/override:

> Sorbet only does these checks if you opt-in using abstract and override, which is one reason I think people might not realize they're there. (One reason for that behavior is that, because Ruby has one massive class hierarchy, people sometimes end up "overriding" parent methods in a way that's not supposed to actually be an override and is therefore entirely incompatible with the type of the parent method. A good example here is Socket#send, which is for sending data and has nothing at all to do with Object#send, which is a metaprogramming-ey way of calling methods: if Sorbet strictly enforced co/contravariance in all places, then this stdlib type would be in flagrant violation of those rules.)

I think that at the very least, Sorbet should be able to figure out whether a runtime exception would be raised (e.g., the parent method is defined in source code, not in the standard library).

At best, we might be able to also make it apply to standard library definitions, and write an autocorrect to insert `allow_incompatible: true` to the bad overrides:

```ruby
# typed: true
require 'sorbet-runtime'
class Parent
  extend T::Sig
  extend T::Helpers
  abstract!
  sig {abstract.returns(Integer)}
  def foo
  end
end
class Child < Parent
  extend T::Sig
  sig {override(allow_incompatible: true).returns(String)}
  def foo
    ''
  end
end
p Parent.new.foo
p Child.new.foo
```
[→ View on sorbet.run](https://sorbet.run/#%23%20typed%3A%20true%0Arequire%20'sorbet-runtime'%0Aclass%20Parent%0A%20%20extend%20T%3A%3ASig%0A%20%20extend%20T%3A%3AHelpers%0A%20%20abstract!%0A%20%20sig%20%7Babstract.returns(Integer)%7D%0A%20%20def%20foo%0A%20%20end%0Aend%0Aclass%20Child%20%3C%20Parent%0A%20%20extend%20T%3A%3ASig%0A%20%20sig%20%7Boverride(allow_incompatible%3A%20true).returns(String)%7D%0A%20%20def%20foo%0A%20%20%20%20''%0A%20%20end%0Aend%0Ap%20Parent.new.foo%0Ap%20Child.new.foo)

**Comment by jez** [comment]
Some more links for the sake of testing

1. [→ View on sorbet.run](https://sorbet.run/#%23%20typed%3A%20true%0Aclass%20Parent%0A%20%20extend%20T%3A%3ASig%0A%0A%20%20sig%20%7Boverridable.returns(Integer)%7D%0A%20%20def%20declared_overridable%0A%20%20%20%200%0A%20%20end%0A%0A%20%20sig%20%7Breturns(Integer)%7D%0A%20%20def%20not_declared_overridable%0A%20%20%20%200%0A%20%20end%0Aend%0A%0Aclass%20Child%20%3C%20Parent%0A%20%20extend%20T%3A%3ASig%0A%0A%20%20sig%20%7Boverride.returns(T.any(String%2C%20Integer))%7D%0A%20%20def%20declared_overridable%0A%20%20%20%20''%0A%20%20end%0A%0A%20%20sig%20%7Boverride.returns(T.any(String%2C%20Integer))%7D%0A%20%20def%20not_declared_overridable%0A%20%20%20%20''%0A%20%20end%0Aend%0A%0Abegin%0A%20%20p%20Parent.new.declared_overridable%0A%20%20p%20Child.new.declared_overridable%0Arescue%20RuntimeError%0Aend%0A%0Abegin%0A%20%20p%20Parent.new.not_declared_overridable%0A%20%20p%20Child.new.not_declared_overridable%0Arescue%20RuntimeError%0Aend)

1. [→ View on sorbet.run](https://sorbet.run/#%23%20typed%3A%20true%0Aclass%20Parent%0A%20%20extend%20T%3A%3ASig%0A%0A%20%20sig%20%7Boverridable.returns(T.any(String%2C%20Integer))%7D%0A%20%20def%20declared_overridable%0A%20%20%20%200%0A%20%20end%0A%0A%20%20sig%20%7Breturns(T.any(String%2C%20Integer))%7D%0A%20%20def%20not_declared_overridable%0A%20%20%20%200%0A%20%20end%0Aend%0A%0Aclass%20Child%20%3C%20Parent%0A%20%20extend%20T%3A%3ASig%0A%0A%20%20sig%20%7Boverride.returns(String)%7D%0A%20%20def%20declared_overridable%0A%20%20%20%20''%0A%20%20end%0A%0A%20%20sig%20%7Boverride.returns(String)%7D%0A%20%20def%20not_declared_overridable%0A%20%20%20%20''%0A%20%20end%0Aend%0A%0Abegin%0A%20%20p%20Parent.new.declared_overridable%0A%20%20p%20Child.new.declared_overridable%0Arescue%20RuntimeError%0Aend%0A%0Abegin%0A%20%20p%20Parent.new.not_declared_overridable%0A%20%20p%20Child.new.not_declared_overridable%0Arescue%20RuntimeError%0Aend)

1. [→ View on sorbet.run](https://sorbet.run/#%23%20typed%3A%20true%0Aclass%20Parent%0A%20%20extend%20T%3A%3ASig%0A%0A%20%20sig%20%7Boverridable.returns(Integer)%7D%0A%20%20def%20declared_overridable%0A%20%20%20%200%0A%20%20end%0A%0A%20%20sig%20%7Breturns(Integer)%7D%0A%20%20def%20not_declared_overridable%0A%20%20%20%200%0A%20%20end%0Aend%0A%0Aclass%20Child%20%3C%20Parent%0A%20%20extend%20T%3A%3ASig%0A%0A%20%20sig%20%7Boverride.returns(String)%7D%0A%20%20def%20declared_overridable%0A%20%20%20%20''%0A%20%20end%0A%0A%20%20sig%20%7Boverride.returns(String)%7D%0A%20%20def%20not_declared_overridable%0A%20%20%20%20''%0A%20%20end%0Aend%0A%0Abegin%0A%20%20p%20Parent.new.declared_overridable%0A%20%20p%20Child.new.declared_overridable%0Arescue%20RuntimeError%0Aend%0A%0Abegin%0A%20%20p%20Parent.new.not_declared_overridable%0A%20%20p%20Child.new.not_declared_overridable%0Arescue%20RuntimeError%0Aend)

**Comment by jez** [comment]
Fixed in #3710 and #3808
